NIFI-1857: HTTPS Site-to-Site

ijokarumawak opened the pull request:
Hi,

I've squashed #423 with the latest master, to add proxy auth and better connection management. However, I haven't done tests with the latest clustering model which doesn't use NCM.

- Enable HTTP(S) for Site-to-Site communication
- Support HTTP Proxy in the middle of local and remote NiFi
- Support BASIC and DIGEST auth with Proxy Server
- Provide 2-phase style commit same as existing socket version
- [WIP] Test with the latest cluster env (without NCM) hasn't tested yet

Details are also written in [NiFi Feature Proposals](https://cwiki.apache.org/confluence/x/v628Aw).

ijokarumawak commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-resources/src/main/resources/conf/nifi.properties:
```diff
@@ -111,9 +111,11 @@ nifi.components.status.repository.buffer.size=${nifi.components.status.repositor
+nifi.remote.input.host=
 nifi.remote.input.secure=true
+nifi.remote.input.socket.port=
+nifi.remote.input.http.enabled=true
```
`nifi.remote.input.http.enabled` and `nifi.remote.input.secure` are set to true by default, but https port, keystore and truststore are not configured. With the default nifi.properties file, user will see following exception at start up, which is not a good UX. Can we change default `nifi.remote.input.secure` to `false` ?

```
Caused by: java.lang.RuntimeException: Remote input HTTPS is enabled but nifi.web.https.port is not specified.
        at org.apache.nifi.util.NiFiProperties.getRemoteInputHttpPort(NiFiProperties.java:439) ~[nifi-properties-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
```

ijokarumawak commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/remote/StandardRemoteProcessGroup.java:
```diff
@@ -1117,31 +1164,39 @@ public void run() {
                     pointsToCluster.set(isPointingToCluster);
-                } else if (statusCode == UNAUTHORIZED_STATUS_CODE) {
-                    try {
-                        final ClientResponse requestAccountResponse = utils.issueRegistrationRequest(apiUri.toString());
```
Hi @mcgilman @markap14 
RPG used to create a NiFi user automatically by sending a request to `/controller/users` which UserResource listens to. But that resource has been removed by #359 and it's not listed on [Support Authorizer API](https://cwiki.apache.org/confluence/display/NIFI/Support+Authorizer+API) resource list.
Does that mean we no longer have to create an user automatically?

ijokarumawak commented:
@mcgilman @markap14 I haven't touched RemoteProcessGroupAuditor.java with this PR since I couldn't figure out the usage of this class. Processors shows its audit in NiFi UI, but where can I see audits for RemoteProcessGroup?

ijokarumawak commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-site-to-site/src/main/java/org/apache/nifi/remote/StandardRootGroupPort.java:
```diff
@@ -236,7 +235,8 @@ private void onTrigger(final ProcessContext context, final ProcessSession sessio
 
-        session.commit();
+        // TODO: Comfirm this. Session.commit here is not required since it has been committed inside receiveFlowFiles/transferFlowFiles.
+        // session.commit();
```
HTTP Site-to-Site has to split the transferring process to two HTTP requests. The first request calls StandardRootGroupPort's onTrigger() as Socket Site-to-Site does. But HTTP version has to hold the transaction, and the second HTTP request actually does the session.commit. Whether it calls `session.commit()` is determined by each transport protocol inside `receiveFlowFiles` or `transferFlowFiles`.

This `session.commit` doesn't let HTTP version to hold the transaction, and it was redundant since it has been committed inside receiveFlowFiles/transferFlowFiles.
I just wanted to confirm that this doesn't break anything. Would you confirm it @markap14 ?

ijokarumawak commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/SiteToSiteResource.java:
```diff
@@ -86,12 +156,958 @@ public Response getController() {
+
+        // TODO: Followed how SocketRemoteSiteListener define peerUrl and clusterUrl, but it can be more meaningful values, especially for clusterUrl.
+        String peerUrl = "nifi://" + clientHostName + ":" + clientPort;
+        String clusterUrl = "nifi://localhost:" + req.getLocalPort();
```
What can be a meaningful clusterUrl here?

markap14 commented:
@ijokarumawak I can't seem to get this to work properly. I configured nifi to allow http site-to-site only, not secure. I then connected a GenerateFlowFile processor to a Remote Process Group that points back to http://localhost:8080/nifi with the only port that I have. I ensured that the port was running and transferring data to an UpdateAttribute processor that ends the flow. When I start sending data to the RPG, though, no data is sent, and the logs show the following every 60 seconds:

```
2016-06-07 12:14:44,902 WARN [I/O dispatcher 89] o.a.n.r.util.SiteToSiteRestApiClient Didn't received additional packet to send, nor confirm transaction call for 30001 millis which exceeds idle connection expiration millis(30000). This transaction will timeout.
2016-06-07 12:14:44,902 ERROR [I/O dispatcher 89] o.a.n.r.util.SiteToSiteRestApiClient Sending data to http://127.0.0.1:8080/nifi-api/site-to-site/input-ports/84cd8e54-68ba-41ff-a1a3-ffc445836a8f/transactions/dddd06ca-3912-4a8f-9e81-fff68d8c0e96/flow-files has failed
java.io.IOException: Didn't received additional packet to send, nor confirm transaction call for 30001 millis
	at org.apache.nifi.remote.util.SiteToSiteRestApiClient$3.produceContent(SiteToSiteRestApiClient.java:504) ~[nifi-site-to-site-client-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.http.impl.nio.client.MainClientExec.produceContent(MainClientExec.java:262) ~[httpasyncclient-4.1.1.jar:4.1.1]
	at org.apache.http.impl.nio.client.DefaultClientExchangeHandlerImpl.produceContent(DefaultClientExchangeHandlerImpl.java:136) ~[httpasyncclient-4.1.1.jar:4.1.1]
	at org.apache.http.nio.protocol.HttpAsyncRequestExecutor.outputReady(HttpAsyncRequestExecutor.java:240) ~[httpcore-nio-4.4.4.jar:4.4.4]
	at org.apache.http.impl.nio.DefaultNHttpClientConnection.produceOutput(DefaultNHttpClientConnection.java:292) ~[httpcore-nio-4.4.4.jar:4.4.4]
	at org.apache.http.impl.nio.client.InternalIODispatch.onOutputReady(InternalIODispatch.java:86) [httpasyncclient-4.1.1.jar:4.1.1]
	at org.apache.http.impl.nio.client.InternalIODispatch.onOutputReady(InternalIODispatch.java:39) [httpasyncclient-4.1.1.jar:4.1.1]
	at org.apache.http.impl.nio.reactor.AbstractIODispatch.outputReady(AbstractIODispatch.java:147) [httpcore-nio-4.4.4.jar:4.4.4]
	at org.apache.http.impl.nio.reactor.BaseIOReactor.writable(BaseIOReactor.java:190) [httpcore-nio-4.4.4.jar:4.4.4]
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent(AbstractIOReactor.java:343) [httpcore-nio-4.4.4.jar:4.4.4]
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents(AbstractIOReactor.java:317) [httpcore-nio-4.4.4.jar:4.4.4]
	at org.apache.http.impl.nio.reactor.AbstractIOReactor.execute(AbstractIOReactor.java:278) [httpcore-nio-4.4.4.jar:4.4.4]
	at org.apache.http.impl.nio.reactor.BaseIOReactor.execute(BaseIOReactor.java:106) [httpcore-nio-4.4.4.jar:4.4.4]
	at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run(AbstractMultiworkerIOReactor.java:590) [httpcore-nio-4.4.4.jar:4.4.4]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_60]
```

ijokarumawak commented:
@markap14 Thanks for taking time to try, and sorry to hear that it didn't work. I'll check it. Thanks!

ijokarumawak commented:
@markap14 I tested sending data using HTTP Site-to-Site with fresh build of this PR. It successfully sent data. The exception happens when a client sends data packet, but doesn't send additional data packet nor confirm the transaction. It seems Remote Process Group Port stop working in the middle..

I've uploaded my [test flow template and nifi.properties](https://gist.github.com/ijokarumawak/03b781e5dea6e706421fdfabf4b30c8d). The flow should be the same as you mentioned. Would you take a look and let me know if there's any difference.

ijokarumawak commented:
@markap14 I think I was able to reproduce the issue, it seems it happens when theres more data in the queue than a RemoteProcessGroup can send at once. I'll fix the issue.

ijokarumawak commented:
@markap14 Fixed the sending data issue. The reason was I didn't call ByteBuffer clear() and didn't closed PipedOutputStream properly. Sorry for the inconvenience. Please try it again.

Also, I fixed the Remote Process Group Port so that useCompression can be set from UI. It seems the UI configuration hasn't been used. I'll send another PR for 0.x branch for that.

markap14 commented:
@ijokarumawak i checked out the new PR and tried the test again. Updated nifi.properties only to set secure = false for site-to-site (I think this needs to be the default because as-is, nifi doesn't startup out of the box).

I did get further this time, and saw the receiving side trying to receive data but still got Exceptions and no data coming through. The logs show:

```
2016-06-08 10:39:04,660 ERROR [Timer-Driven Process Thread-10] o.a.nifi.remote.StandardRemoteGroupPort RemoteGroupPort[name=Log,target=http://localhost:8080/nifi] failed to communicate with remote NiFi instance due to java.io.IOException: Failed to confirm transaction with Peer[url=http://127.0.0.1:8080/nifi-api] due to java.io.IOException: Unexpected response code: 500 errCode:Abort errMessage:Server encountered an exception.
2016-06-08 10:39:05,668 ERROR [NiFi Web Server-24] o.apache.nifi.web.api.SiteToSiteResource Unexpected exception occurred. clientId=a252a4c6-5a5f-42f3-8270-322923b8c118, portId=30638675-e655-4719-b684-905ad0d49eac
2016-06-08 10:39:05,671 ERROR [NiFi Web Server-24] o.apache.nifi.web.api.SiteToSiteResource Exception detail:
org.apache.nifi.processor.exception.ProcessException: org.apache.nifi.processor.exception.FlowFileAccessException: Failed to import data from org.apache.nifi.stream.io.MinimumLengthInputStream@75b15a7e for StandardFlowFileRecord[uuid=d8046125-80be-4475-bb43-b15cf6dec4d8,claim=,offset=0,name=531446421738053,size=0] due to org.apache.nifi.processor.exception.FlowFileAccessException: Unable to create ContentClaim due to java.io.EOFException
	at org.apache.nifi.remote.StandardRootGroupPort.receiveFlowFiles(StandardRootGroupPort.java:503) ~[nifi-site-to-site-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.web.api.SiteToSiteResource.receiveFlowFiles(SiteToSiteResource.java:418) ~[classes/:na]
	at sun.reflect.GeneratedMethodAccessor348.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_60]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_60]
	at com.sun.jersey.spi.container.JavaMethodInvokerFactory$1.invoke(JavaMethodInvokerFactory.java:60) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.model.method.dispatch.AbstractResourceMethodDispatchProvider$ResponseOutInvoker._dispatch(AbstractResourceMethodDispatchProvider.java:205) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.model.method.dispatch.ResourceJavaMethodDispatcher.dispatch(ResourceJavaMethodDispatcher.java:75) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.HttpMethodRule.accept(HttpMethodRule.java:302) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.ResourceClassRule.accept(ResourceClassRule.java:108) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.uri.rules.RootResourceClassesRule.accept(RootResourceClassesRule.java:84) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1542) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1473) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409) [jersey-server-1.19.jar:1.19]
	at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409) [jersey-servlet-1.19.jar:1.19]
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:558) [jersey-servlet-1.19.jar:1.19]
	at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:733) [jersey-servlet-1.19.jar:1.19]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) [javax.servlet-api-3.1.0.jar:3.1.0]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:845) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1689) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlets.GzipFilter.doFilter(GzipFilter.java:51) [jetty-servlets-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.apache.nifi.web.filter.RequestLogger.doFilter(RequestLogger.java:66) [classes/:na]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:316) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:126) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:90) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:122) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.apache.nifi.web.security.NiFiAuthenticationFilter.authenticate(NiFiAuthenticationFilter.java:82) [nifi-web-security-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.web.security.NiFiAuthenticationFilter.doFilter(NiFiAuthenticationFilter.java:56) [nifi-web-security-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.apache.nifi.web.security.NiFiAuthenticationFilter.authenticate(NiFiAuthenticationFilter.java:82) [nifi-web-security-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.web.security.NiFiAuthenticationFilter.doFilter(NiFiAuthenticationFilter.java:56) [nifi-web-security-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.apache.nifi.web.security.NiFiAuthenticationFilter.authenticate(NiFiAuthenticationFilter.java:82) [nifi-web-security-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.web.security.NiFiAuthenticationFilter.doFilter(NiFiAuthenticationFilter.java:56) [nifi-web-security-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.apache.nifi.web.security.node.NodeAuthorizedUserFilter.doFilter(NodeAuthorizedUserFilter.java:113) [nifi-web-security-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:213) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:176) [spring-security-web-4.0.3.RELEASE.jar:4.0.3.RELEASE]
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:346) [spring-web-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:262) [spring-web-4.2.4.RELEASE.jar:4.2.4.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1676) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.apache.nifi.web.filter.TimerFilter.doFilter(TimerFilter.java:51) [classes/:na]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1668) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:581) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548) [jetty-security-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1174) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:511) [jetty-servlet-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1106) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:119) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.Server.handle(Server.java:524) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:319) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:253) [jetty-server-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273) [jetty-io-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95) [jetty-io-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93) [jetty-io-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589) [jetty-util-9.3.9.v20160517.jar:9.3.9.v20160517]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_60]
Caused by: org.apache.nifi.processor.exception.FlowFileAccessException: Failed to import data from org.apache.nifi.stream.io.MinimumLengthInputStream@75b15a7e for StandardFlowFileRecord[uuid=d8046125-80be-4475-bb43-b15cf6dec4d8,claim=,offset=0,name=531446421738053,size=0] due to org.apache.nifi.processor.exception.FlowFileAccessException: Unable to create ContentClaim due to java.io.EOFException
	at org.apache.nifi.controller.repository.StandardProcessSession.importFrom(StandardProcessSession.java:2325) ~[nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.remote.protocol.AbstractFlowFileServerProtocol.receiveFlowFiles(AbstractFlowFileServerProtocol.java:441) ~[nifi-site-to-site-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.remote.StandardRootGroupPort.receiveFlowFiles(StandardRootGroupPort.java:248) ~[nifi-site-to-site-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.remote.StandardRootGroupPort.onTrigger(StandardRootGroupPort.java:222) ~[nifi-site-to-site-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.remote.StandardRootGroupPort.onTrigger(StandardRootGroupPort.java:157) ~[nifi-site-to-site-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunConnectableTask.call(ContinuallyRunConnectableTask.java:81) ~[nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunConnectableTask.call(ContinuallyRunConnectableTask.java:40) ~[nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:123) ~[nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_60]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) ~[na:1.8.0_60]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) ~[na:1.8.0_60]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) ~[na:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_60]
	... 1 common frames omitted
Caused by: org.apache.nifi.processor.exception.FlowFileAccessException: Unable to create ContentClaim due to java.io.EOFException
	at org.apache.nifi.controller.repository.StandardProcessSession.importFrom(StandardProcessSession.java:2318) ~[nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	... 14 common frames omitted
Caused by: java.io.EOFException: null
	at org.apache.nifi.stream.io.MinimumLengthInputStream.read(MinimumLengthInputStream.java:60) ~[nifi-utils-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.stream.io.MinimumLengthInputStream.read(MinimumLengthInputStream.java:53) ~[nifi-utils-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.stream.io.StreamUtils.copy(StreamUtils.java:35) ~[nifi-utils-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.repository.FileSystemRepository.importFrom(FileSystemRepository.java:699) ~[nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.apache.nifi.controller.repository.StandardProcessSession.importFrom(StandardProcessSession.java:2315) ~[nifi-framework-core-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	... 14 common frames omitted
```

ijokarumawak commented:
@markap14 I'm not 100% sure if the change will solve the problem since I couldn't reproduce the issue myself.

However, I've added additional buffer draining code after receiving EOF from channel.
In addition to that, a failure detection code was added to check If it fails sending data as much as expected, when it does, it throws RuntimeException so that we can investigate further.

Also, I've changed the default nifi.remote.input.secure to `false`.

Could you give it a try once again? If it sill fails, please share your flow template. Thank you.

markap14 commented:
@ijokarumawak this is looking good now! I have pulled down the latest PR, rebased against master, and have been able to test this running directly against my NiFi instance and while using nginx as a proxy. Nicely done! I have pushed this to master.

markap14 commented:
@ijokarumawak I added the "This closes #497" message to the commit that I pushed, but it doesn't seem to have worked... can you close the PR?

ijokarumawak commented:
@markap14 Thanks for taking your time to review and test, I'm so glad to hear that worked!
The CI test haven't been finished, maybe that's why this isn't closed yet. I'll check the status again later.
